eo-phi-normalizer/src/Language/EO/Phi/Rules/Common.hs

```diff
@@ -202,6 +205,42 @@ intToBytes n = Bytes $ insertDashes $ pad $ showHex n ""
               (x : y : xs) -> x : y : '-' : go xs
          in go s
 
+minNu :: Int
+minNu = -1
+
+class HasMaxNu a where
+  -- | get maximum vertex index
+  --
+  -- >>> getMaxNu @Object "⟦ a ↦ ⟦ ν ↦ ⟦ Δ ⤍ 03- ⟧ ⟧, b ↦ ⟦ ⟧ ⟧"
+  -- 3
+  getMaxNu :: a -> Int
+
+instance HasMaxNu Program where
+  getMaxNu :: Program -> Int
+  getMaxNu (Program bindings) = maximum (minNu : (getMaxNu <$> bindings))
+
+instance HasMaxNu Object where
+  getMaxNu :: Object -> Int
+  getMaxNu = \case
+    Formation bindings -> maximum (minNu : (getMaxNu <$> bindings))
+    Application obj bindings -> max (getMaxNu obj) (maximum (minNu : (getMaxNu <$> bindings)))
+    ObjectDispatch obj _ -> getMaxNu obj
+    _ -> minNu
+
+instance HasMaxNu Binding where
+  getMaxNu :: Binding -> Int
+  getMaxNu = \case
+    AlphaBinding VTX (Formation [DeltaBinding (Bytes bs)]) ->
+      case readHex [x | x <- bs, x /= '-'] of
+        [(val, "")] -> val
+        _ -> error "Vertex number is incorrect"
+    AlphaBinding _ obj -> getMaxNu obj
+    _ -> minNu
+
+instance HasMaxNu Attribute where
+  getMaxNu :: Attribute -> Int
+  getMaxNu = const 0
```

Comment:
Is this instance really needed? Does it make sense to query `Nu`s on an attribute?
Also, why is it 0 and not `minNu`?